### DISCUSSION TITLE
A `data-wp-input` directive for two-way data binding

### DISCUSSION THREAD

**Discussion by DAreRodz** [WordPress/gutenberg#62332] [Interactivity API]
As per @sirreal suggestion, having a directive that eases the control of `<input>` elements would be interesting, so it is not necessary to write a callback to handle value updates.

The directive could just point to a property in the state or context:

```html
<input data-wp-input="state.value">
```

No event handler would be necessary to write something like this:

```html
<form
    data-wp-interactive="example/input"
    data-wp-context="{}"
>
    <input
        type="text"
        data-wp-input="context.text"
    >
    <output
        data-wp-text="context.text"
    ></output>
</form>
```

It could even handle input types so you don't need to parse the value:

```html
<script>
    import { store, getContext } from "@wordpress/interactivity";

    store( 'example/input', {
        state: {
            get result() {
                const { a, b } = getContext();
                return a + b;
            }
        }
    } );
</script>

<form
    data-wp-interactive="example/input"
    data-wp-context='{ "a": 0, "b": 0 }'
>
    <input
        name="A"
        type="number"
        data-wp-input="context.a"
    >
    <input
        name="B"
        type="number"
        data-wp-input="context.b"
    >
    <output
        name="A + B"
        data-wp-text="state.result"
    ></output>
</form>
```

The directive is mostly intended to be used in `<input>` elements (maybe `<textarea>` as well), so I thought about naming it `data-wp-input`. We can choose other names, though. 😄

Another name that comes to my mind is `data-wp-control`, as that's the word used in the [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input) to describe `<input>` elements.

Let me know what you think. 😊

**Comment by sirreal** [comment]
Thanks for creating this! It does't provide anything that cannot already be implemented, but it would save some repetitive code. It might also prevent bugs because change/input handlers would not need to be implemented for everything to connect to the store.

One way to think about it is as the inverse of the `bind` directives. `data-wp-bind--value="…"` gets a value from the store or context and sets an element attribute. This directive would take an element attribute and update a value in the store or context: `data-wp-input--value="…"`.

The most common use case would likely be for `<input>` but I wonder if any element and any attribute could theoretically be read (`checked`, `value`, `textContent`). Maybe a more generic directive name would be good like `data-wp-read--[attribute]` or `data-wp-inject--[attribute]` although I don't love either of those names 🙂

**Comment by cbravobernal** [comment]
How would be an example of integrating input validation with those directives?
With `data-wp-run`?

**Reply by sirreal** [reply]
I think it would be OK to scope this to basic functionality of get value and put it in state/context. More advanced usage could continue to use events and handler functions.

**Comment by luisherranz** [comment]
I like it 😄👏

I like that it is different from `data-wp-bind` so that you can do one-way binding or two-way binding depending on the case.

**Reply by singerb** [reply]
As long as two way binding with two directives worked and didn't infinite loop, I think that's fine.

**Comment by singerb** [comment]
+1 from me to this idea; I had to read the API docs twice to make sure I wasn't missing what seemed like a standard feature for frameworks like this. Having to write my own code for it was fine with a single form element, but with anything more complex it would get repetitive very quickly.

**Comment by vitalikpetreniuk** [comment]
What about this feature, will it be implemented?

**Reply by sirreal** [reply]
There doesn't seem to be opposition, so with sufficient interest I expect it would be. Then it becomes a question of priorities.

If someone from the community would like to work on it that would be welcome!

**Comment by nickchomey** [comment]
I'm very glad to discover this discussion/proposal!

 I am a long-time WP user, but have just started using the iAPI after a couple years of using Datastar and a bit of Alpine on other projects. Datastar's [`bind` attribute](https://data-star.dev/reference/attributes#data-bind) and Alpine's [model directive](https://alpinejs.dev/directives/model) work as the proposed input directive would.

I will start working on an input directive that brings this much-needed functionality. 

ps.  I think `bind` would actually be a better name for this and then rename the existing `bind` to `attr` (as Datastar does), but we're already stuck with bind being used to set attributes

**Reply by nickchomey** [reply]
I've submitted a PR here: https://github.com/WordPress/gutenberg/pull/80021

Would be great to get a review!
